tRecorderApi/api/views/user.py

```diff
@@ -1,8 +1,166 @@
+from drf_yasg import openapi
+from drf_yasg.utils import swagger_auto_schema
+from rest_framework import status
+from rest_framework import views
 from rest_framework import viewsets
-from api.serializers import UserSerializer
+from rest_framework.authentication import TokenAuthentication
+from rest_framework.authtoken.models import Token
+from rest_framework.parsers import JSONParser
+from rest_framework.permissions import IsAuthenticatedOrReadOnly
+from rest_framework.response import Response
+
+from api.permissions import CanCreateOrDestroyOrReadonly
+from ..models.user import User
+from ..serializers import UserSerializer
 
+
+@method_decorator(name='list', decorator=swagger_auto_schema(
+    operation_description="Return list of users based on given query string",
+    manual_parameters=[
+        openapi.Parameter(
+            name='id', in_=openapi.IN_QUERY,
+            type=openapi.TYPE_INTEGER,
+            description="Id of a user",
+        ), openapi.Parameter(
+            name='icon_hash', in_=openapi.IN_QUERY,
+            type=openapi.TYPE_STRING,
+            description="Icon hash of a user",
+        ), openapi.Parameter(
+            name='is_social', in_=openapi.IN_QUERY,
+            type=openapi.TYPE_BOOLEAN,
+            description="Social status of a user. Whether a user was created via social media or identicon.",
+        )
+    ]
+))
 class UserViewSet(viewsets.ModelViewSet):
     """This class handles the http GET, PUT, PATCH, POST and DELETE requests."""
     queryset = User.objects.all()
-    serializer_class = UserSerializer
\ No newline at end of file
+    serializer_class = UserSerializer
+    permission_classes = (CanCreateOrDestroyOrReadonly,)
+
+    def get_queryset(self):
+        queryset = []
+        query = self.request.query_params
+        if len(query) == 0:
+            return User.objects.filter(is_superuser=False)
+        else:
+            pk = query.get("id", None)
+            icon_hash = query.get("icon_hash", None)
+            is_staff = query.get("is_staff", None)
+            if pk is not None:
+                queryset = User.objects.filter(id=pk, is_superuser=False)
+            if icon_hash is not None:
+                queryset = User.objects.filter(icon_hash=icon_hash, is_superuser=False)
+            if is_staff is not None:
+                if is_staff == "true":
+                    queryset = User.objects.filter(is_staff=True, is_superuser=False)
+                else:
+                    queryset = User.objects.filter(is_staff=False, is_superuser=False)
+
+            if len(queryset) != 0:
+                return queryset
+            else:
+                return None
+
+    def destroy(self, request, pk=None):
+        instance = self.get_object()
+        try:
+            os.remove(instance.name_audio)
+        except OSError:
+            pass
+        self.perform_destroy(instance)
+        return Response(status=status.HTTP_200_OK)
+
+    def create(self, request):
```

Comment:
This method is doing a lot that looks like it should be broken down into smaller methods, maybe something like the following:

if created:
validate_profile_upload() // verifies we have all the blobs/content needed
create_profile() // converts audio, writes content to disk at appropriate location
add_user() //creates a user object once all necessary files are stored and in their proper place

I didn't spend much time thinking about the names, just trying to communicate the couple distinct operations I see. Makes all of these distinct operations testable (and they should get test)

Reply:
Fixed in latest commit